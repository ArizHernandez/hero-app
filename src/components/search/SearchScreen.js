import React, { useMemo } from 'react'
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroes/HeroCard';
import queryString from 'query-string';
import { getHerosByName } from '../../selectors/getHerosByName';

export const SearchScreen = ({ history }) => {

  const location = useLocation();
  const { hero = '' } = queryString.parse( location.search );

  const [ { heroName }, handleInputChange ] = useForm({ 
    heroName: hero
  });
  
  const heroFilter = useMemo(() => getHerosByName({ name: hero }), [ hero ]);

  const handleSearch = (e) => {
    e.preventDefault();

    if( heroName === '' ){ return; }

    history.push(`?hero=${ heroName }`);
  }


  return (
    <div className="container">
      <h1 className="mt-3 text-center">Search Heros</h1>
      <hr className="mt-0"/>

      <div className="row">
        <div className="col-12 col-md-5 col-lg-4">
          <h3 className="text-center text-md-start">Search</h3>
          <form onSubmit={handleSearch}>
            <div className="input-group">
              <input 
                className="form-control"
                type="search"
                name="heroName"
                placeholder="Hero Name"
                value={heroName}
                onChange={handleInputChange}
              />
              <button
                className="btn btn-primary"
              >
                Search
              </button>
            </div>
          </form>
        </div>

        <div className="col">
          <h3 className="text-center mt-4 mt-md-0">Results</h3>
          <div className="row">
              {
                hero === '' 
                ?
                  <p className="mt-3 text-center fs-4 text-muted"> Search a new hero, now!! </p>
                  :
                  heroFilter.length === 0 
                  ? 
                  <p className="mt-3 text-center fs-4 alert alert-danger"> Hero not found </p>
                  :
                    heroFilter.map( hero => {
                      return (
                        <div 
                          key={ hero.id }
                          className="col-8 col-md-5 mb-2 mx-auto"
                        >
                          <HeroCard
                            { ...hero }
                          />
                        </div>
                      )
                    })
              }
          </div>
        </div>
      </div>
    </div>
  )
}
