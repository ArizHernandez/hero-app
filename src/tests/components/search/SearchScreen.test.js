import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en <SearchScreen />', () => {
  
  test('Debe de cargarse correctamente', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route
          path='/search'
          component={SearchScreen}
        />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.text-muted').text().trim()).toBe('Search a new hero, now!!');
  });
  
  test('Debe de mostrar a batman y el input con el valor del query string', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?hero=batman']}>
        <Route
          path='/search'
          component={SearchScreen}
        />
      </MemoryRouter>
    );
    
    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });
  
  test('Debe de mostrar un mensaje de error si el heroe buscado no existe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?hero=test-error']}>
        <Route  
          path='/search'
          component={SearchScreen}
        />
      </MemoryRouter>
    );

    expect(wrapper.find('.alert-danger').text().trim()).toBe('Hero not found');
  });

  test('Debe de llamar el push del history', () => {
    const historyMock = {
      push: jest.fn()
    }

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path='/search'
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target:{
        name:'heroName', value:'batman'
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });

    expect(historyMock.push).toBeCalledWith('?hero=batman')
  });
})
