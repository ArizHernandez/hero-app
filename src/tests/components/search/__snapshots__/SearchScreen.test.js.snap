// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de cargarse correctamente 1`] = `
<div
  className="container"
>
  <h1
    className="mt-3 text-center"
  >
    Search Heros
  </h1>
  <hr
    className="mt-0"
  />
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5 col-lg-4"
    >
      <h3
        className="text-center text-md-start"
      >
        Search
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="heroName"
            onChange={[Function]}
            placeholder="Hero Name"
            type="search"
            value=""
          />
          <button
            className="btn btn-primary"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col"
    >
      <h3
        className="text-center mt-4 mt-md-0"
      >
        Results
      </h3>
      <div
        className="row"
      >
        <p
          className="mt-3 text-center fs-4 text-muted"
        >
           Search a new hero, now!! 
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar a batman y el input con el valor del query string 1`] = `
<div
  className="container"
>
  <h1
    className="mt-3 text-center"
  >
    Search Heros
  </h1>
  <hr
    className="mt-0"
  />
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5 col-lg-4"
    >
      <h3
        className="text-center text-md-start"
      >
        Search
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <input
            className="form-control"
            name="heroName"
            onChange={[Function]}
            placeholder="Hero Name"
            type="search"
            value="batman"
          />
          <button
            className="btn btn-primary"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col"
    >
      <h3
        className="text-center mt-4 mt-md-0"
      >
        Results
      </h3>
      <div
        className="row"
      >
        <div
          className="col-8 col-md-5 mb-2 mx-auto"
          key="dc-batman"
        >
          <div
            className="card text-center"
          >
            <img
              alt="..."
              className="card-img-top"
              src=""
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text text-end mb-2"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                className="btn btn-dark btn-block"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
